### PR TITLE
Implement remote.Write

### PR DESCRIPTION
**Pull request by mattmoor**
Fixes: #15

### CODE REVIEW

**Comment by mattmoor** [comment]
Based on: https://github.com/google/go-containerregistry/pull/40

File: v1/remote/write.go [hunk 1]
```diff
@@ -34,5 +38,227 @@ type WriteOptions struct {
...
+		options: wo,
+	}
+
+	bs, err := img.BlobSet()
+	if err != nil {
+		return err
+	}
+
+	// Spin up go routines to publish each of the members of BlobSet(),
+	// and use an error channel to collect there results.
```
Comment (jonjohnsonjr) [inline v1/remote/write.go]: "their results"

File: v1/remote/write.go [hunk 2]
```diff
@@ -34,5 +38,227 @@ type WriteOptions struct {
...
+	if err != nil {
+		return err
+	}
+
+	if err := w.commitBlob(h, location); err != nil {
+		return err
+	}
+	log.Printf("pushed %v", h)
+	return nil
+
```
Comment (jonjohnsonjr) [inline v1/remote/write.go]: extra newline here

File: v1/remote/write.go [hunk 3]
```diff
@@ -34,5 +38,227 @@ type WriteOptions struct {
...
+	}
+
+	// Now wait for all of the blob uploads to complete.
+	var errors []error
+	for _ = range bs {
+		if err := <-errCh; err != nil {
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
```
Comment (dlorenc) [inline v1/remote/write.go]: This feels a bit weird. Maybe append all the errors together and return a combined one?

https://github.com/hashicorp/go-multierror for inspiration.

File: v1/remote/write.go (hunk 1 above)
Comment (mattmoor) [reply v1/remote/write.go]: Done.

File: v1/remote/write.go [hunk 4]
```diff
@@ -34,5 +38,227 @@ type WriteOptions struct {
...
+		Scheme: transport.Scheme(w.ref.Context().Registry),
+		Host:   w.ref.Context().RegistryStr(),
+		Path:   path,
+	}
+}
+
+// nextLocation extracts the fully-qualified URL to which we should send the next request in an upload sequence.
+func (w *writer) nextLocation(resp *http.Response) string {
+	loc := resp.Header.Get("Location")
+	if len(loc) > 0 && loc[0] == '/' {
```
[the remark this answers is not in the document]
Comment (mattmoor) [reply v1/remote/write.go]: Done.

File: v1/remote/write.go [hunk 5]
```diff
@@ -34,5 +38,227 @@ type WriteOptions struct {
...
+		// how to commit this layer.
+		return w.nextLocation(resp), nil
+	default:
+		return "", fmt.Errorf("unrecognized status code during PATCH: %v", resp.Status)
+	}
+}
+
+// commitBlob commits this blob by sending a PUT to the location returned from streaming the blob.
+func (w *writer) commitBlob(h v1.Hash, location string) (err error) {
+	req, err := http.NewRequest(http.MethodPut, location, nil)
```
[the remark this answers is not in the document]
Comment (mattmoor) [reply v1/remote/write.go]: Done.

File: v1/remote/write.go (hunk 2 above)
Comment (mattmoor) [reply v1/remote/write.go]: Done.

File: v1/remote/write.go (hunk 3 above)
Comment (mattmoor) [reply v1/remote/write.go]: I only wait for all of them to avoid writes on a closed channel :P

**Pull request merged**